### PR TITLE
[Arch] Streamline EventStream Runtime Image Building Logic

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

The previous logic of runtime build is complex to wrap one's head around (e.g., if you don't set `update_source_code=True` while developing, you might be scratching your head wondering what's going on, and it turns out you used an outdated image).

Furthermore, the `update_source_code=True` is always set in most `evaluation/` (#3230) to ensure that the evaluation always reflects the current code. In #3243, we try to use directory hash to track the source code and `Dockerfile` used to build each sandbox image and hopefully avoid repeat rebuilding for evaluation.

However, I found the existing logic still requires an additional round of re-build for evaluation instances if `update_source_code=True`:
1. The first time running of SWE-Bench instance docker image "AAA:T" with source code hash ("abc") will build "od_runtime:AAA_tag_T", which is also tagged to "od_runtime:abc"
2. Assume the source code is unchanged, the second time running of SWE-Bench instance, it will detect "od_runtime:AAA_tag_T" exists, and since `update_source_code=True`, it will skip checking the hash, and rebuild an "od_runtime_dev:AAA_tag_T".

After drinking a bit, carefully maintaining my BAC level, and reaching the Ballmer's Peak, I realized the `update_source_code` might not be necessary at all.

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR removes `update_source_code`, and re-write the logic to keep track of runtime images:

Given a user-specified arbitrary docker image (e.g., `ubuntu:22.04`), it will be converted to an OpenDevin-supported Docker image.

The resulting image will be tagged as both:
- `{target_image_repo}:{target_image_hash_tag}`
- `{target_image_repo}:{target_image_tag}`

Explanation:
- `target_image_hash_tag` is the hash of the docker build folder (containing source code and Dockerfile).
    And same hash means the image is built from scratch with the *exact SAME source code* on the
    exact Dockerfile (i.e., same base image name & version).
- `target_image_tag` is the generic tag for the target image that's generic and resembles the "latest" tag
    in the docker image naming convention for a particular {repo}:{tag} pair.
    For example, od_runtime:od_v0.8.3_ubuntu_tag_22.04 means the latest OpenDevin v0.8.3 image built from scratch on top of
    ubuntu:22.04.

How does the image building works:
1. We convert the base image (e.g., `ubuntu:22.04`) into od_runtime image name (e.g., `od_runtime:od_v0.8.3_ubuntu_tag_22.04`)
2. We create a docker build folder (assuming we build from scratch), generate Dockerfile and copy over source code, and calculate a hash `from_scratch_hash`.
3. If the image exists with the same hash (e.g., `od_runtime:{from_scratch_hash}`), we will not build any image, and will pull that image and use it as it.
4. If the exact hash is not found (e.g., `od_runtime:{from_scratch_hash}`), but a recent OD runtime of the same image exists (e.g., `od_runtime:od_v0.8.3_ubuntu_tag_22.04`), we will try to rebuild the new image (e.g., `od_runtime:{from_scratch_hash}`) based on the latest image (e.g., `od_runtime:od_v0.8.3_ubuntu_tag_22.04`) to save some dependency installation time.
6. If the image hash is not found (e.g., `od_runtime:{from_scratch_hash}`) AND we cannot re-use the latest relavant image (e.g., `od_runtime:od_v0.8.3_ubuntu_tag_22.04`), we will build it completely from scratch.

---
**Other references**

### CODE REVIEW

File: opendevin/runtime/utils/runtime_build.py [hunk 1]
```diff
@@ -12,6 +12,10 @@
 import opendevin
 from opendevin.core.logger import opendevin_logger as logger
 
+RUNTIME_IMAGE_REPO = os.getenv(
+    'OD_RUNTIME_RUNTIME_IMAGE_REPO', 'ghcr.io/opendevin/od_runtime'
```
Comment (tobitege) [inline opendevin/runtime/utils/runtime_build.py:16]: isn't 1 RUNTIME enough? 😁

Comment (xingyaoww) [reply opendevin/runtime/utils/runtime_build.py:16]: Yeah, I'd ideally want to keep all the flexibility here (e.g., We might use a different image registry for the hosted version :D)

File: docs/modules/usage/runtime.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+sidebar_position: 4
+---
+
+# OpenDevin EventStream Runtime
+
+The OpenDevin EventStream Runtime is the core component that enables secure and flexible execution of AI agent's action.
+It creates a sandboxed environment using Docker, where arbitrary code can be run safely without risking the host system.
+
+
+## Why we need sandboxed runtime?
```
Comment (neubig) [inline docs/modules/usage/runtime.md:11]:
```suggestion
## Why do we need a sandboxed runtime?
```

File: docs/modules/usage/runtime.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+- It executes various types of actions (shell commands, file operations, Python code, etc.) safely within the container.
+- It manages the state of the sandboxed environment, including the current working directory and loaded plugins.
+- It formats and returns observations to the backend, ensuring a consistent interface for processing results.
+
+
+## Advanced: How OpenDevin builds and maintains OD Runtime images
+
+OpenDevin uses a sophisticated approach to build and manage runtime images. This process ensures efficiency, consistency, and flexibility in creating and maintaining Docker images for both production and development environments.
+
+Checkout [relavant code](https://github.com/OpenDevin/OpenDevin/blob/main/opendevin/runtime/utils/runtime_build.py) if you are interested in more details.
```
Comment (neubig) [inline docs/modules/usage/runtime.md:84]:
```suggestion
Check out [relavant code](https://github.com/OpenDevin/OpenDevin/blob/main/opendevin/runtime/utils/runtime_build.py) if you are interested in more details.
```

**Review by neubig — APPROVED** [review]
🍻

**Pull request merged**